### PR TITLE
Build: Remove custom bundler and extract minification

### PR DESCRIPTION
**Pull request by samouri**
This PR Seeks to remove our custom bundler from the build process so that we can use `webpack` via cli.

*why*
- simplify.  using a custom bundling script opens the door for straying from the straight and narrow which can make upgrade paths more difficult.
- scope-hoisting/tree shaking: while implementing the [pr](https://github.com/Automattic/wp-calypso/pull/16057) to enable these features, it was difficult to enable es6 within the bundler.  By removing the bundler entirely, we won't need to worry about as many untranspiled bits in the process
- speed: the bundler had only been being run as a `postbuild` step, but with this change the client/server can be built in parallel
- Wouldn't it be great if we could build our client just by running: `webpack --config webpack.config.js`

Implementation details:
- no more assets middleware.
- once all of the client js files are generated,  minification is done via a separate focused script that can be run via `npm run minify`

Note: This is a branch off of https://github.com/Automattic/wp-calypso/pull/16379.  It has essentially the same properties except extracts the minification process so that we can keep it parallelized because waiting for 1.0 of webpack-uglify-plugin is proving to take too long

### CODE REVIEW

**Comment by samouri** [comment]
cc @ehg who i'd been telling I wanted to do this for months

File: server/pages/index.jade [hunk 1]
```diff
@@ -152,18 +152,11 @@ html(lang=lang, dir=isRTL ? 'rtl' : 'ltr', class=isFluidWidth ? 'is-fluid-width'
...
-			script(src=urls[ jsFile ])
-			if chunk
-				script(src=urls[ chunk ])
-		else
-			script(src=urls[ 'manifest-min' ])
-			script(src=urls[ 'vendor-min' ])
-			script(src=urls[ jsFile + '-min' ])
-			if chunk
-				script(src=urls[ chunk + '-min' ])
+		script(src=urls[ 'manifest' ])
```
Comment (blowery) [inline server/pages/index.jade]: any chance we can inline this now?

Comment (samouri) [reply server/pages/index.jade]: Not a bad idea, i was thinking about doing that.  Ok, I'll do it.

Comment (samouri) [reply server/pages/index.jade]: @blowery: do you mind if we do that in a future PR?

Comment (blowery) [reply server/pages/index.jade]: don't mind

File: server/pages/index.jade [hunk 2]
```diff
@@ -152,18 +152,11 @@ html(lang=lang, dir=isRTL ? 'rtl' : 'ltr', class=isFluidWidth ? 'is-fluid-width'
 			script(type='text/javascript')!= config
 		if i18nLocaleScript
 			script(src=i18nLocaleScript)
-		if 'development' === env || isDebug
-			script(src=urls[ 'manifest' ])
-			script(src=urls[ 'vendor' ])
-			script(src=urls[ jsFile ])
-			if chunk
-				script(src=urls[ chunk ])
-		else
```
Comment (blowery) [inline server/pages/index.jade]: without out this if / else how do we tell prod to serve minified files?

Comment (samouri) [reply server/pages/index.jade]: nice catch! i need to revert this file now that we have separate `min` files again.

File: server/pages/index.js [hunk 3]
```diff
@@ -104,15 +116,8 @@ function generateStaticUrls( request ) {
 		urls[ file.path ] = getUrl( file.path, file.hash );
 	} );
 
-	const assets = request.app.get( 'assets' );
-
-	assets.forEach( function( asset ) {
-		const name = asset.name;
-		urls[ name ] = asset.url;
-		if ( config( 'env' ) !== 'development' ) {
```
Comment (blowery) [inline server/pages/index.js]: need this bit too to generate the `-min` variants to hand off to the jade template

Comment (samouri) [reply server/pages/index.js]: eh. it feels dumb to need that logic in two places.  I'll just place it here and not have the jade template care.

**Comment by samouri** [comment]
Just tested:
- [x] docker build: only minified files were served. no errors seen
- [x] local dev: only unminified files were served.

File: package.json [hunk 4]
```diff
@@ -283,6 +284,7 @@
     "socket.io": "1.4.5",
     "stylelint": "6.9.0",
     "supertest": "2.0.0",
+    "uglify-js": "2.6.4",
```
Comment (blowery) [inline package.json]: why'd we do back in uglify versions? was `2.7.0`

**Comment by blowery** [comment]
@ehg we're waiting on https://github.com/webpack-contrib/uglifyjs-webpack-plugin/pull/108 and https://github.com/webpack-contrib/uglifyjs-webpack-plugin/pull/121 to land I think.

**Comment by blowery** [comment]
```
❯ ./node_modules/.bin/webpack --config webpack.config.js
module.js:471
    throw err;
    ^

Error: Cannot find module 'assets-webpack-plugin'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/blowery/src/a8c/wp-calypso/webpack.config.js:16:22)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at requireConfig (/Users/blowery/src/a8c/wp-calypso/node_modules/webpack/bin/convert-argv.js:97:18)
    at /Users/blowery/src/a8c/wp-calypso/node_modules/webpack/bin/convert-argv.js:104:17
    at Array.forEach (native)
    at module.exports (/Users/blowery/src/a8c/wp-calypso/node_modules/webpack/bin/convert-argv.js:102:15)
    at Object.<anonymous> (/Users/blowery/src/a8c/wp-calypso/node_modules/webpack/bin/webpack.js:155:40)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
```

Should that work?

**Comment by blowery** [comment]
should this work?

```
❯ ./node_modules/.bin/webpack --config webpack.config.js
module.js:471
    throw err;
    ^

Error: Cannot find module 'lib/create-config'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/blowery/src/a8c/wp-calypso/server/config/index.js:6:22)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/blowery/src/a8c/wp-calypso/webpack.config.js:22:16)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at requireConfig (/Users/blowery/src/a8c/wp-calypso/node_modules/webpack/bin/convert-argv.js:97:18)
    at /Users/blowery/src/a8c/wp-calypso/node_modules/webpack/bin/convert-argv.js:104:17
```

**Comment by blowery** [comment]
`NODE_PATH=server:client ./node_modules/.bin/webpack --config webpack.config.js` does work

**Comment by blowery** [comment]
Overall seems fine. I'd like to know why we dropped back the uglifyjs version before merging.

**Comment by samouri** [comment]
> Overall seems fine. I'd like to know why we dropped back the uglifyjs version before merging.

total accident. fixing!

**Comment by samouri** [comment]
We use NODE_PATHs a lot which makes for pretty scary implicit behaviors in our code.
I know this is an issue that has been bugging @gziolo when trying to improve our jest configuration.

**Comment by samouri** [comment]
@ehg that was my initial intention but they've taken longer than expected to get the 1.0 out with parallelization + cache.  It should significantly speed things up and also make our code even simpler.

It'll be a very easy transition once `1.0` does come out!

**Comment by samouri** [comment]
giving this a final round of testing before hitting merge

**Comment by gziolo** [comment]
I’m glad this one has landed 👏 How far are we towards adding .babel.js suffix to both Webpack configs? 😁 I think this one is a big step forward. It is very important improvement if we want to run codemods without worrying that we break some code running with Node. Even if we migrate to Node 8 we need to remember that we are using Babel setup with many features that aren’t part of the official JS standard.

Related question is if we should start using node-babel to run scripts from bin folder. Those two ingredients would allow to use unified syntax everywhere in the Calypso codebase. 💥

**Comment by samouri** [comment]
> How far are we towards adding .babel.js suffix to both Webpack configs? 
not far at all :).  Expect a PR coming soon

**Comment by jsnajdr** [comment]
@samouri After this PR, `stats.json` file is no longer being saved when `WEBPACK_OUTPUT_JSON=1` is set as environment variable. The part of the `bundler.js` script that did the saving is gone.

This breaks the `npm run analyze-bundles` feature and also iscalypsofastyet.com is unable to analyze the recent builds 😢 

Can we get it back? :wink:

**Comment by Tug** [comment]
It seems to have broken `webpack-dev-middleware` as well, I don't have source-maps in dev anymore. Investigating...

File: webpack.config.js [hunk 5]
```diff
@@ -70,7 +71,7 @@ const jsLoader = isWindows ? babelLoader : 'happypack/loader';
 const webpackConfig = {
 	bail: calypsoEnv !== 'development',
 	entry: {},
-	devtool: '#eval',
+	devtool: 'false',
```
Comment (Tug) [inline webpack.config.js]: Why disable `devtool` here?

**Pull request merged**